macaw-aarch32-symbolic/src/Data/Macaw/AArch32/Symbolic/ABI.hs

```diff
@@ -0,0 +1,184 @@
+{-|
+Copyright        : (c) Galois, Inc 2024
+Maintainer       : Langston Barrett <langston@galois.com>
+
+On AArch32, the stack grows \"downwards\" from high addresses to low. The end
+of the stack is initialized with the ELF auxiliary vector (not modelled here),
+and functions expect the following data to be available above their stack frame
+(i.e., just above the address in @sp@/@r13@), from high addresses to low:
+
+* Padding (if necessary)
+* Their stack-spilled arguments
+
+The stack is always 4-byte aligned, and must be 8-byte aligned \"at any public
+interface\", i.e., at function calls.
+
+Unlike x86_64, the return address is not stored on the stack.
+
+The following diagram summarizes the stack frame layout:
+
+@
+High addresses
+
+|---------------------|
+| Caller's frame      |
+|---------------------|
+| Padding (if needed) |
+|---------------------|
+| Spilled argument n  |
+|---------------------|
+| ...                 |
+|---------------------|
+| Spilled argument 2  |
+|---------------------|
+| Spilled argument 1  |
+|---------------------| <- sp
+| Callee's frame      |
+|---------------------|
+
+Low addresses
+@
+
+The functions in this module support manipulation of a stack under these
+constraints. ABI-compatible machine code translated by Macaw manages the stack
+for itself, so this module is primarily helpful for initial setup of the stack,
+before starting symbolic execution.
+
+Helpful links:
+
+* <https://github.com/ARM-software/abi-aa/blob/a82eef0433556b30539c0d4463768d9feb8cfd0b/aapcs32/aapcs32.rst#621the-stack>
+* <https://learn.microsoft.com/en-us/cpp/build/overview-of-arm-abi-conventions?view=msvc-170>
+
+This module is meant to be imported qualified.
+-}
+
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE ImplicitParams #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+
+module Data.Macaw.AArch32.Symbolic.ABI
+  ( StackPointer
+  , getStackPointer
+  , stackPointerReg
+  , alignStackPointer
+  , SpilledArgs(..)
+  , writeSpilledArgs
+  , pushStackFrame
+  , allocStack
+  ) where
+
+import Control.Lens qualified as Lens
+import Data.Coerce (coerce)
+import Data.Macaw.AArch32.Symbolic qualified as AArch32S
+import Data.Macaw.ARM qualified as AArch32
+import Data.Macaw.Symbolic qualified as MS
+import Data.Macaw.Symbolic.Stack qualified as MSS
+import Data.Parameterized.Classes (ixF')
+import Data.Parameterized.Context qualified as Ctx
+import Data.Sequence qualified as Seq
+import Lang.Crucible.Backend qualified as C
+import Lang.Crucible.LLVM.DataLayout qualified as CLD
+import Lang.Crucible.LLVM.MemModel qualified as MM
+import Lang.Crucible.Simulator qualified as C
+import What4.Interface qualified as WI
+
+-- | Helper, not exported
+ptrRepr :: WI.NatRepr 32
+ptrRepr = WI.knownNat
+
+-- | A pointer to an AArch32 stack
+newtype StackPointer sym = StackPointer { getStackPointer :: MM.LLVMPtr sym 32 }
```

Comment:
We now have multiple data types named `StackPointer`: one in `macaw-aarch32-symbolic` and another in `macaw-x86-symbolic`. They are similar enough that I wonder if we shouldn't just factor these out into `macaw-symbolic`:

```hs
newtype StackPointer sym w = StackPointer { getStackPointer :: MM.LLVMPtr sym w }
```

Do you agree? (Similar comments apply to `SpilledArgs` and `RetAddr`.)

Reply:
I actually think it's nice to have these as separate types, even though they have similar representations. They both contain pointers, but those pointers have different constraints, and having separate types prevents confusion between them. For example, it would be wrong to pass a pointer returned from `Data.Macaw.AArch32.Symbolic.ABI.stackPointerReg` into `Data.Macaw.X86.Symbolic.ABI.SysV.pushStackFrame`, because the ABI constraints for AArch32 and SysV are different. Perhaps more realistically, consider what would happen if we added support for the Microsoft x86_64 ABI as well. Then, we wouldn't want to confuse pointers to a Microsoft-compatible stack with pointers to a SysV-compatible stack.